include/libtorrent/aux_/file_view_pool.hpp

```diff
@@ -30,87 +30,124 @@ POSSIBILITY OF SUCH DAMAGE.
-		// the overload that takes ``file_index`` releases only the file with
-		// that index in storage ``st``.
+		file_view open_file(storage_index_t st, std::string const& p
+			, file_index_t file_index, file_storage const& fs, open_mode_t m);
+
+		// release all file views belonging to the specified storage_interface
+		// (``st``) the overload that takes ``file_index`` releases only the file
+		// with that index in storage ``st``.
 		void release();
 		void release(storage_index_t st);
 		void release(storage_index_t st, file_index_t file_index);
 
 		// update the allowed number of open file handles to ``size``.
 		void resize(int size);
 
-		// returns the current limit of number of allowed open file handles held
-		// by the file_pool.
+		// returns the current limit of number of allowed open file views held
+		// by the file_view_pool.
 		int size_limit() const { return m_size; }
 
-		// internal
-		void set_low_prio_io(bool b) { m_low_prio_io = b; }
 		std::vector<open_file_state> get_status(storage_index_t st) const;
 
-		// close the file that was opened least recently (i.e. not *accessed*
-		// least recently). The purpose is to make the OS (really just windows)
-		// clear and flush its disk cache associated with this file. We don't want
-		// any file to stay open for too long, allowing the disk cache to accrue.
-		void close_oldest();
-
 	private:
 
-		file_handle remove_oldest(std::unique_lock<std::mutex>&);
+		std::shared_ptr<file_mapping> remove_oldest(std::unique_lock<std::mutex>&);
 
 		int m_size;
-		bool m_low_prio_io = false;
 
-		struct lru_file_entry
+		using file_id = std::pair<storage_index_t, file_index_t>;
+
+		struct file_entry
 		{
-			file_handle file_ptr;
-			time_point const opened{aux::time_now()};
-			time_point last_use{opened};
+			file_entry(file_id k
+				, string_view name
+				, open_mode_t const m
+				, std::int64_t const size)
+				: key(k)
+				, mapping(std::make_shared<file_mapping>(file_handle(name, size, m), m, size))
+				, mode(m)
+			{}
+
+			file_id key;
+			std::shared_ptr<file_mapping> mapping;
+			time_point last_use{aux::time_now()};
 			open_mode_t mode{};
 		};
 
-		// maps storage pointer, file index pairs to the
-		// LRU entry for the file
-		std::map<std::pair<storage_index_t, file_index_t>, lru_file_entry> m_files;
+		using files_container = mi::multi_index_container<
+			file_entry,
+			mi::indexed_by<
+			// look up files by (torrent, file) key
+			mi::ordered_unique<mi::member<file_entry, file_id, &file_entry::key>>,
+			// look up files by least recently used
+			mi::sequenced<>
+			>
+		>;
+
+		// maps storage pointer, file index pairs to the lru entry for the file
+		files_container m_files;
 		mutable std::mutex m_mutex;
+
+		// the boost.multi-index container is not no-throw move constructable. In
+		// order to destruct m_files without holding the mutex, we need this
+		// separate pre-allocated container to move it into before releasing the
+		// mutex and clearing it.
```

Comment:
I don't understand this comment. What problem does moving `m_files` then clearing the moved-to container avoid, and what does the move ctor not being no-throw have to do with it? It looks to me like this code is just trying to avoid holding the main mutex while doing the potentially long-running job of closing all of the mappings.

It's also not clear to me why `m_deferred_destruction` needs to be a member instead of a local variable inside of `file_view_pool::release()`.

Reply:
(from memory): You're right that the ultimate goal is to avoid holding the mutex while destructing files, as it may be expensive (especially on MacOS/HFS+). To achieve this, I move the main container (`m_files`) into another, temporary, one. release the mutex and then destruct the temporary container. Traditionally this is done by just constructing a local container, on the stack, and move the member container into that one.

The problem with that is that (default) constructing that local temporary container may throw an exception. I imagine this happens at a point in the program where correctly handling errors makes things very complex, so instead I pre-allocate the temporary container here, to always have it available. If the default construction will throw, it at least happens at a better time, when the file_view_pool is being constructed.

Reply:
I suppose one could argue that locking the mutex also may fail, and I'm not handling those errors correctly. I don't have those instrumented in my simulation that triggers exceptions though.